snuba/clickhouse/astquery.py

```diff
@@ -118,23 +120,37 @@ def _format_query_impl(self) -> str:
         if self.__limit is not None:
             limit_clause = f"LIMIT {self.__limit} OFFSET {self.__offset}"
 
-        self.__formatted_query = " ".join(
-            [
-                c
-                for c in [
-                    select_clause,
-                    from_clause,
-                    array_join_clause,
-                    prewhere_clause,
-                    where_clause,
-                    group_clause,
-                    having_clause,
-                    order_clause,
-                    limitby_clause,
-                    limit_clause,
-                ]
-                if c
+        self.__sql_data_list = [
+            (k, v)
+            for k, v in [
+                ("select", select_clause),
+                ("from", from_clause),
+                ("join", array_join_clause),
+                ("prewhere", prewhere_clause),
+                ("where", where_clause),
+                ("group", group_clause),
+                ("having", having_clause),
+                ("order", order_clause),
+                ("limitby", limitby_clause),
+                ("limit", limit_clause),
             ]
-        )
+            if v
+        ]
+
+        return self.__sql_data_list
+
+    def _format_query_impl(self) -> str:
+        if self.__formatted_query:
+            return self.__formatted_query
+
+        self.__formatted_query = " ".join([c for _, c in self._sql_data_list()])
 
         return self.__formatted_query
+
+    def sql_data(self) -> Dict[str, str]:
```

Comment:
fyi, the typing module gives us some more flexible type hints for collections than just dict, list and tuple.
For example we have Mapping and MutableMapping, Mapping, Sequence and Set are immutable.